src/code/fault_ootn64.c

```diff
@@ -0,0 +1,835 @@
+            for (x = 82; x < 290; x += 52) {
+                Fault_DrawText(x, y, "%08x", *(ptr++));
+            }
+        }
+    }
+}
+
+void Fault_DrawMemDumpPC(OSThread* arg0) {
+    u32* pc = (u32*)arg0->context.pc;
+
+    if ((pc >= (u32*)0x80000020) && (pc <= (u32*)0x80400000)) {
+        Fault_DrawMemDumpContents("PC Trace", pc - 0x20, 0);
+    }
+}
+
+void Fault_DrawMemDumpSP(OSThread* arg0) {
+    u32* sp = (u32*)(u32)arg0->context.sp;
+
+    if ((sp >= (u32*)0x80000020) && (sp <= (u32*)0x80400000)) {
+        Fault_DrawMemDumpContents("Stack Trace", sp, 0);
+    }
+}
+
+void func_800AF3DC(void) {
+    s32 i;
+
+    Fault_DrawRecBlack(22, 16, 276, 208);
+    Fault_DrawText(36, 18, "ROM DEBUG");
+
+    for (i = 0; i < ARRAY_COUNT(B_80122570); i++) {
+        Fault_DrawText(((i % 4) * 52) + 82, ((i / 4) * 9) + 28, "%08x", B_80122570[i]);
+    }
+}
+
+void Fault_ResumeThread(OSThread* thread) {
+    thread->context.cause = 0;
+    thread->context.fpcsr = 0;
+    thread->context.pc += 4;
+    *(u32*)thread->context.pc = 0x0000000D;
+    osWritebackDCache((void*)thread->context.pc, 4);
+    osInvalICache((void*)thread->context.pc, 4);
+    osStartThread(thread);
+}
+
+void func_800AF558(void) {
+    osViBlack(false);
+    if (osViGetCurrentFramebuffer() >= (void*)0x80100000) {
+        gFaultMgr.framebuffer = (s32)osViGetCurrentFramebuffer();
+    } else {
+        gFaultMgr.framebuffer = (s32)(void*)(PHYS_TO_K0(osMemSize) - sizeof(u16[SCREEN_HEIGHT][SCREEN_WIDTH]));
+        osViSwapBuffer((void*)gFaultMgr.framebuffer);
+    }
+}
+
+void Fault_AddClient(FaultClient* client, void* callback, void* arg0, void* arg1) {
+    FaultMgr* faultMgr = &gFaultMgr;
+    OSIntMask mask;
+
+    mask = osSetIntMask(OS_IM_NONE);
+
+    client->callback = callback;
+    client->arg0 = arg0;
+    client->arg1 = arg1;
+    client->next = faultMgr->clients;
+    faultMgr->clients = client;
+
+    osSetIntMask(mask);
+}
+
+void Fault_RemoveClient(FaultClient* client) {
+    u32 mask;
+    FaultClient* iterClient = gFaultMgr.clients;
+    FaultClient* lastClient = NULL;
+
+    mask = osSetIntMask(1);
+    while (iterClient != NULL) {
+        if (iterClient == client) {
+            if (lastClient != NULL) {
+                lastClient->next = client->next;
+            } else {
+                gFaultMgr.clients = client->next;
+            }
+            break;
+        } else {
+            lastClient = iterClient;
+            iterClient = iterClient->next;
+        }
+    }
+    osSetIntMask(mask);
+}
+
+void Fault_ProcessClients(void) {
+    s32 temp_a3;
+    s32 i;
+    FaultClient* client;
+
+    client = gFaultMgr.clients;
+    i = 0;
+    while (client != NULL) {
+        if (client->callback != NULL) {
+            Fault_DrawRecBlack(22, 16, 276, 208);
+            Fault_DrawText(30, 20, "CallBack (%d) %08x %08x %08x", i++, client, client->arg0, client->arg1);
+            ((void (*)(void*, void*))client->callback)(client->arg0, client->arg1);
```

Comment:
why is `FaultClient.callback` a `void*` instead of a function pointer anyway?

Reply:
Good question that also applies to the gc fault handler, I don't know.

Reply:
Maybe add a TODO to investigate later? (Unless someone here knows)

Reply:
@thar0 may know

Reply:
It's because they're massively inconsistent with the function signatures that are assigned to these callbacks, iirc

Reply:
fwiw, MM has these as distinct typedefs and just have `void*` casts in `Fault_AddClient` calls where needed. `AddrConv` clients seem consistent.

Reply:
Alright, I think this is fine for now personally